Improve readme of repository (#3647)
FAForever/fa: setup/readme.md

We define the `bin` directory to be the `bin` folder in the installation folder of the client. By default this is:
 - `C:\ProgramData\FAForever\bin`

Copy the content of `repository/setup/bin` into the `bin` folder. Open up `init_dev.lua` that now resides in the `bin` folder. At the top it states:

```lua
-- change this to the location of the repository on your disk. Note that `\` is used
-- for escaping characters and you can just use `/` in your path instead.
local locationOfRepository = 'your-fa-repository-location'
```

Change that to match the path to the repository on your system. You can start the game by calling `start_dev.bat` or `start_dev.sh`. You can inspect them to find out what they do.

Attaching the debugger
----------------------

With thanks to KionX we have a [debugger](https://github.com/FAForever/FADeepProbe). When an exception occurs it can trace the exception to a line of Lua code. You can download it via the releases. Store the executable in your `bin` folder. This is the same `bin` folder as defined earlier. We need to adapt the bat / bash files to use the debugger. As an example we change the bat file from:

```bat

> ForgedAlliance.exe /init "init_dev.lua" /EnableDiskWatch /showlog /log "dev.log"

Reviewer: FADeepProbe.exe?

Author: I don't understand 😄

Reviewer: Right now the first argument is the FA exe instead of the debugger exe on this line

Author: Yes - we need to change it to the debugger. See the line below.

Reviewer: Oh my bad when I first looked at this I thought one was bat example and the other was sh example like the rest so my bad reading.

```
To:
```bat
FADeepProbe.exe /init "init_dev.lua" /EnableDiskWatch /showlog /log "dev.log"
```

The arguments are passed along by the debugger. The change to the bash script is similar. When the game crashes the debugger will try and inform you in the log what happened.

Work environment
----------------

> When you're a novel programmer then you may not have a work environment. This is often referred to as an Integrated Development Environment (IDE). We recommend you to use [Visual Studio Code](https://code.visualstudio.com/) (VSC)

Reviewer 2: What? Novel is not the right word here.

Author: Replaced it with 'new' .

Other useful information:
 - [Shortcuts](https://code.visualstudio.com/shortcuts/keyboard-shortcuts-windows.pdf) for Visual Studio Code

Shortcuts can significantly improve your workflow. You do not need to familiarize yourself with each and every shortcut. We recommend you to look at this sheet every month and see if there is a common action that you've been doing that can be done by a shortcut instead. 

Useful extensions in general:
 - [Git graph](https://marketplace.visualstudio.com/items?itemName=mhutchie.git-graph): useful for graph visualizations
 - [Gitlens](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens): useful for seeing who made what change
 - [Peacock](https://marketplace.visualstudio.com/items?itemName=johnpapa.vscode-peacock): useful for coloring your workspaces when you have multiple

Useful extensions if you intent to work with shaders:
 - [Shader languages support for VS Code](https://marketplace.visualstudio.com/items?itemName=slevesque.shader)
 - [HLSL Tools](https://marketplace.visualstudio.com/items?itemName=TimGJones.hlsltools)
 - [Hex editor](https://marketplace.visualstudio.com/items?itemName=ms-vscode.hexeditor)

> Note that I do not recommend the Lua language server. The game uses a slightly adjusted version of Lua. The syntax doesn't match. We also have a different import system. The Lua language server generates dozens of errors. And it is not equipped to work with our import system.

Reviewer 2: Might be able to use faf-lua with it. I have it set up with Atom on linux, although I haven’t compiled faf-lua for windows.

Author: That would be interesting but I'm not sure how to do that myself

Instructions to set up your development environment
---------------------------------------------------

> A collection of useful fragments of information to help you set up your development environment.

Reviewer: ```suggestion
A collection of useful information to help you set up your development environment.
```

Running the game with your changes
----------------------------------

_There is a section about Git in the FAQ if you're unfamilar with it._

> Fork the repository. Clone your fork to your system using your favorite Git tool. We refer to the `repository` directory to the location of your repository on your system. We refer to the `bin` directory to the `bin` folder in the installation folder of the client. By default this is:

Reviewer: ```suggestion
Fork the repository. Clone your fork to your system using your favorite Git tool. We refer to the `repository` directory as the location of your repository on your system. We refer to the `bin` directory as the `bin` folder in the installation folder of the client. By default this is:
```

 - `C:/ProgramData/FAForever/bin`

> Copy the content of `repository/setup/bin` into the `bin` folder. Open up `init_dev.lua` that now resides in the `bin` folder. At the top it states:

Reviewer: ```suggestion
Copy the contents of `repository/setup/bin` into the `bin` folder. Open `init_dev.lua` now found in the `bin` folder. At the top it states:
```

 - `%userprofile%\AppData\Local\Gas Powered Games\Supreme Commander Forged Alliance\Game.prefs`

Hooking
-------

All files in the `schook` directory hook files in the `lua` folder. A mod applies a similar hooking strategy. Hooking means concatenating files: one file is appended to the next. This allows you to add code to a file. As an example, say we have this base game file:
 - `lua/sim/unit.lua`

And say we are trying to hook it with both:
 - `schook/lua/unit.lua`
 - `mods/my-mod/hook/lua/unit.lua`

Then what happens is:
 - `lua/sim/unit.lua` = `lua/sim/unit.lua` + `schook/lua/unit.lua` + `mods/my-mod/hook/lua/unit.lua`

Where the addition operator should be interpreted as appending one text (source) file to the other text (source) file. this extents the unit file. This hooking is done when a file is imported for the first time. The directory that is used for hooks can be configured in the initialization files.

Frequently asked Questions (FAQ)
--------------------------------

There is more specific information on the [wiki](https://github.com/FAForever/fa/wiki) about the structure of the repository or how a release works.

<dl>
<dt> What is Git? </dt>

> <dd> Git is software for tracking changes in a set of files. This is particular useful for source files. It is an industry standard. It is a good time investment to familiarize yourself with it. There are various sources to learn git. You can learn using Git via the <a href="https://www.w3schools.com/git/git_intro.asp?remote=github">command line</a>, via <a href="https://desktop.github.com/">Github Desktop</a>[Github Desktop](https://desktop.github.com/) or via <a href="https://code.visualstudio.com/docs/introvideos/versioncontrol">Visual Studio Code</a>. At first it is important to understand what a fork is, how to stage, commit or push your files and how you can make a pull request. More advanced topics like merging stratgies or rebasing get relevant as you become more involved with the project. </dd>

Reviewer 3: Un-rendered markdown `[Github Desktop](https://desktop.github.com/)` included in between html links.

An asside, if you include a blank line after a html tag it will begin rendering markdown again without affecting the html.
